aptos-move/aptos-vm/src/aptos_vm.rs

```diff
@@ -1438,14 +1440,114 @@ impl AptosVM {
         &self,
-        _resolver: &impl AptosMoveResolver,
-        _txn: ValidatorTransaction,
-        _log_context: &AdapterLogSchema,
-    ) -> (VMStatus, VMOutput) {
-        (
-            VMStatus::Executed,
-            VMOutput::empty_with_status(TransactionStatus::Keep(ExecutionStatus::Success)),
-        )
+        resolver: &impl AptosMoveResolver,
+        txn: ValidatorTransaction,
+        log_context: &AdapterLogSchema,
+    ) -> Result<(VMStatus, VMOutput), VMStatus> {
+        let session_id = SessionId::validator_txn(&txn);
+        match txn {
+            ValidatorTransaction::DKGResult(dkg_node) => {
+                self.process_dkg_result(resolver, log_context, session_id, dkg_node)
+            },
+            ValidatorTransaction::DummyTopic1(_) => Ok((
+                VMStatus::Executed,
+                VMOutput::empty_with_status(TransactionStatus::Keep(ExecutionStatus::Success)),
+            )),
+            ValidatorTransaction::DummyTopic2(_) => Ok((
+                VMStatus::Executed,
+                VMOutput::empty_with_status(TransactionStatus::Keep(ExecutionStatus::Success)),
+            )),
+        }
+    }
+
+    fn process_dkg_result(
+        &self,
+        resolver: &impl AptosMoveResolver,
+        log_context: &AdapterLogSchema,
+        session_id: SessionId,
+        dkg_node: DKGNode,
+    ) -> Result<(VMStatus, VMOutput), VMStatus> {
+        //TODO(zjma): replace `StatusCode::INVALID_SIGNATURE` in this function
+        let dkg_state = load_on_chain_config_from_resolver::<DKGState>(resolver)
+            .map_err(|e| {
+                VMStatus::error(
+                    StatusCode::INVALID_SIGNATURE,
+                    Some(format!(
+                        "process_dkg_result failed with dkg state loading error: {e}"
+                    )),
+                )
+            })?
+            .ok_or_else(|| {
+                VMStatus::error(
+                    StatusCode::INVALID_SIGNATURE,
+                    Some("process_dkg_result failed with missing DKGState resource".to_string()),
+                )
+            })?;
+        let DKGState { in_progress, .. } = dkg_state;
+        let in_progress_session_state = in_progress.ok_or_else(|| {
+            VMStatus::error(
+                StatusCode::INVALID_SIGNATURE,
+                Some(
+                    "process_dkg_result failed with missing in-progress session state".to_string(),
+                ),
+            )
+        })?;
+
+        // Check epoch number.
+        if dkg_node.metadata.epoch != in_progress_session_state.metadata.dealer_epoch {
+            return Err(VMStatus::error(
+                StatusCode::INVALID_SIGNATURE,
+                Some("process_dkg_result failed with invalid epoch".to_string()),
+            ));
+        }
+
+        // Deserialize transcript and verify it.
+        let pub_params = DummyDKG::new_public_params(&in_progress_session_state.metadata);
```

Comment:
why dummydkg here?

Reply:
That's the only DKGTrait implementor today. We can switch to a real one, or rename it to real later.

What's really sad is that DKGTrait doesn't help - here we have to specify a concrete "scheme".